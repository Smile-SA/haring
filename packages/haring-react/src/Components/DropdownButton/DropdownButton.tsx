'use client';

import type { MenuProps } from '@mantine/core';
import type { ReactElement, ReactNode } from 'react';

import { Button, Menu } from '@mantine/core';
import { CaretDown, CaretUp } from '@phosphor-icons/react';
import { useId, useState } from 'react';

export interface IDropdownButtonProps extends MenuProps {
  /** Override default menu target with custom component*/
  buttonComponent?: ReactNode;
  children?: ReactNode;
  label?: string;
  role?: string;
}

/** Additional props will be forwarded to the [Mantine Menu component](https://mantine.dev/core/menu) */
export function DropdownButton(props: IDropdownButtonProps): ReactElement {
  const {
    buttonComponent,
    children,
    label,
    position = 'bottom-start',
    ...menuProps
  } = props;
  const [opened, setOpened] = useState(false);
  const id = useId();

  return (
    <Menu
      id={id}
      onChange={setOpened}
      opened={opened}
      position={position}
      {...menuProps}
    >
      <Menu.Target>
        {buttonComponent ?? (
          <Button
            data-testid="button"
            rightSection={opened ? <CaretUp /> : <CaretDown />}
          >
            {label}
          </Button>
        )}
      </Menu.Target>
      <Menu.Dropdown data-testid="dropdown">{children}</Menu.Dropdown>
    </Menu>
  );
}
