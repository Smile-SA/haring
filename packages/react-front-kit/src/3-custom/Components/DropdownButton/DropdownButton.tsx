'use client';

import type { MenuProps } from '@mantine/core';
import type { ReactNode } from 'react';

import { Button, Menu } from '@mantine/core';
import { CaretDown, CaretUp } from '@phosphor-icons/react';
import { useId, useState } from 'react';

interface IDropdownButtonProps extends MenuProps {
  children?: ReactNode;
  label: string;
}

export function DropdownButton(props: IDropdownButtonProps): JSX.Element {
  const { children, label, position = 'bottom-start', ...menuProps } = props;
  const [opened, setOpened] = useState(false);
  const id = useId();

  return (
    <Menu
      id={id}
      onChange={setOpened}
      opened={opened}
      position={position}
      {...menuProps}
    >
      <Menu.Target>
        <Button
          data-testid="button"
          rightIcon={opened ? <CaretUp /> : <CaretDown />}
        >
          {label}
        </Button>
      </Menu.Target>
      <Menu.Dropdown data-testid="dropdown">{children}</Menu.Dropdown>
    </Menu>
  );
}
