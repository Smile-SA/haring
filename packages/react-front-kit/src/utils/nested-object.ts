export type INestedObjectWithDepth<O extends object> = O & {
  children?: INestedObjectWithDepth<O>[];
  depth?: number;
};

export function flattenNestedObjects<O extends object>(
  arr: INestedObjectWithDepth<O>[],
): INestedObjectWithDepth<O>[] {
  let children: INestedObjectWithDepth<O>[] = [];
  const flattened = arr.map((m) => {
    if (m.children && m.children.length) {
      children = [...children, ...m.children];
    }
    return m;
  });
  return flattened.concat(
    children.length ? flattenNestedObjects(children) : children,
  );
}

export function assignDepth<O extends object>(
  arr?: INestedObjectWithDepth<O>[],
  depth = 0,
): INestedObjectWithDepth<O>[] {
  (arr ?? []).forEach((o) => {
    o.depth = depth;
    assignDepth(o.children, depth + 1);
  });
  return arr ?? [];
}
