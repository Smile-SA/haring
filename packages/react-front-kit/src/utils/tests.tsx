import type {
  Queries,
  RenderOptions,
  RenderResult,
  queries,
} from '@testing-library/react';
import type { ReactElement, ReactNode } from 'react';

import { render } from '@testing-library/react';

import { Provider } from '../3-custom/Provider/Provider';

export function renderWithProviders<
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  ui: ReactElement,
  renderOptions: RenderOptions<Q, Container, BaseElement> = {}
): RenderResult<Q, Container, BaseElement> {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return <Provider>{children}</Provider>;
  }
  return render<Q, Container, BaseElement>(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}
