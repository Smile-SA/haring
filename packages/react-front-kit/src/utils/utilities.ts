export function isNotNullNorEmpty<S>(
  value: Record<string, never> | S,
): value is Exclude<S, null | undefined> {
  return (
    value !== null &&
    value !== undefined &&
    (typeof value === 'object' ? Object.keys(value).length !== 0 : true)
  );
}

export function isCallback<T, U>(maybeFunc: T | U): maybeFunc is T {
  return typeof maybeFunc === 'function';
}
