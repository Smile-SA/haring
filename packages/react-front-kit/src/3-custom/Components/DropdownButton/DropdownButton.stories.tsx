import type { Meta, StoryObj } from '@storybook/react';

import { Menu } from '@mantine/core';
import { expect } from '@storybook/jest';
import { userEvent, within } from '@storybook/testing-library';

import { sleep } from '../../../utils/storybook';

import { DropdownButton as Cmp } from './DropdownButton';

const meta = {
  component: Cmp,
  tags: ['autodocs'],
  title: '3-custom/Components/DropdownButton',
} satisfies Meta<typeof Cmp>;

export default meta;
type IStory = StoryObj<typeof meta>;

export const DropdownButton: IStory = {
  args: {
    children: (
      <>
        <Menu.Item component="a" href="#">
          Calico
        </Menu.Item>
        <Menu.Item component="a" href="#">
          Espace RH
        </Menu.Item>
        <Menu.Item component="a" href="#">
          Aventure IA
        </Menu.Item>
        <Menu.Item component="a" href="#">
          Lunette & CO
        </Menu.Item>
      </>
    ),
    label: 'Mon espace',
  },

  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement.parentElement as HTMLElement);
    await userEvent.click(canvas.getByTestId('button'));
    await sleep(200);
    await expect(canvas.getByTestId('dropdown')).toBeInTheDocument();
  },
};
