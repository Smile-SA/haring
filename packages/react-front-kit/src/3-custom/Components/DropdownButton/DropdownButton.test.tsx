import { renderWithProviders } from '../../../utils/tests';

import { DropdownButton } from './DropdownButton';

describe('DropdownButton', () => {
  it('matches snapshot', () => {
    const { container } = renderWithProviders(<DropdownButton label="label" />);
    expect(container).toMatchSnapshot();
  });
});
