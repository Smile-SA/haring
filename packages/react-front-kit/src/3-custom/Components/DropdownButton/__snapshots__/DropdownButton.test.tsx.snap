// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownButton matches snapshot 1`] = `
<div>
  <button
    aria-controls="mantine-wbb0yg8iz-dropdown"
    aria-expanded="false"
    aria-haspopup="menu"
    class="mantine-UnstyledButton-root mantine-Button-root mantine-xm61av"
    data-button="true"
    id="mantine-wbb0yg8iz-target"
    type="button"
  >
    <div
      class="mantine-1wpc1xj mantine-Button-inner"
    >
      <span
        class="mantine-1ryt1ht mantine-Button-label"
      >
        label
      </span>
      <span
        class="mantine-Button-icon mantine-Button-rightIcon mantine-467wo8"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
          />
        </svg>
      </span>
    </div>
  </button>
</div>
`;
